README.md
[![Community Project header](https://github.com/newrelic/open-source-office/raw/master/examples/categories/images/Community_Project.png)](https://github.com/newrelic/open-source-office/blob/master/examples/categories/index.md#category-community-project)

# nr1-apdex-optimizer

![CI](https://github.com/newrelic/nr1-apdex-optimizer/workflows/CI/badge.svg) ![GitHub release (latest SemVer including pre-releases)](https://img.shields.io/github/v/release/newrelic/nr1-apdex-optimizer?include_prereleases&sort=semver) [![Snyk](https://snyk.io/test/github/newrelic/nr1-apdex-optimizer/badge.svg)](https://snyk.io/test/github/newrelic/nr1-apdex-optimizer)

## Usage

nr1-apdex-optimizer provides actionable recommendations for Apdex threshold settings. These recommendations are based on existing best practices and provide links to the relevant page where you can make the suggested changes.

This Nerdpack lists APM and Browser applications for an account in a table. For each application, you can see the currently configured Apdex thresholds along with links to the application settings page where you can change the Apdex threshold (ApdexT). The table also lists suggested APM and Browser Apdex thresholds, which are updated based on the time picker period.



![Screenshot #1](screenshots/nr1-apdex-optimizer.png)



For great advice on choosing the right Apdex threshold, see Bill Kayser's New Relic [blog post](https://blog.newrelic.com/product-news/how-to-choose-apdex-t/).

### About this Nerdpack

The application relies upon a combination of native NerdGraph and NRQL-based queries. It consists of two top-level React components, the first to identify and list available accounts, and the second to render the table. The table is based on a third-party open source React library, called react-table. It can be sorted on every column, provides pagination, and allows relevant cells to be filtered based on a search term. We have added throughput and errors so you can use those values for prioritization.

## Open source license

This project is distributed under the [Apache 2 license](LICENSE).

## What do you need to make this work?

Make sure you have [New Relic APM](https://newrelic.com/products/application-monitoring) and/or [Browser agents](https://newrelic.com/products/browser-monitoring) installed, as well as related access to [New Relic One](https://newrelic.com/platform).

## Getting started

1. Ensure that you have [Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git) and [NPM](https://www.npmjs.com/get-npm) installed. If you need to check, run the following commands, which will return version numbers if you have one or both installed, and will not be recognized if you don't.

```bash
git --version
npm -v
```

2. Follow the [**Quick start**](https://one.newrelic.com/launcher/developer-center.launcher) instructions to install the NR1 CLI (5 minutes or less) and set up your New Relic development environment.

3. Clone this repo and run the code locally against your New Relic data:

```bash
nr1 nerdpack:clone -r https://github.com/newrelic/nr1-apdex-optimizer.git
cd nr1-apdex-optimizer
nr1 nerdpack:serve
```

Visit [https://one.newrelic.com/?nerdpacks=local](https://one.newrelic.com/?nerdpacks=local), navigate to the Nerdpack, and :sparkles:

## Deploying this Nerdpack

Open a command prompt in the Nerdpack's directory and run the following commands.

```bash
# If you need to create a new uuid for the account to which you're deploying this Nerdpack, use the following
# nr1 nerdpack:uuid -g [--profile=your_profile_name]
# to see a list of APIkeys / profiles available in your development environment, run nr1 credentials:list
nr1 nerdpack:publish [--profile=your_profile_name]
nr1 nerdpack:deploy [-c [DEV|BETA|STABLE]] [--profile=your_profile_name]
nr1 nerdpack:subscribe [-c [DEV|BETA|STABLE]] [--profile=your_profile_name]
```

Visit [https://one.newrelic.com](https://one.newrelic.com), navigate to the Nerdpack and get started.

### Tips for using the app

* If you have an app that has been running for awhile in a steady state and you have a good baseline for acceptable performance, set your Apdex threshold to give you a baseline Apdex score of 0.95. To do this, get the 90th percentile value and set that to Apdex T and set the time picker to 7 days. For information information, see [Changing your Apdex settings](https://docs.newrelic.com/docs/apm/new-relic-apm/apdex/change-your-apdex-settings). 
* When deploying to a master account, select each sub-account from a drop-down menu. If deployed to a single sub-account, the account is automatically selected.

# Support

New Relic has open-sourced this project. This project is provided AS-IS WITHOUT WARRANTY OR DEDICATED SUPPORT. Issues and contributions should be reported here on GitHub.

We encourage you to bring your experiences and questions to the [Explorers Hub](https://discuss.newrelic.com) where our community members collaborate on solutions and new ideas.

## Community

New Relic hosts and moderates an online forum where customers can interact with New Relic employees as well as other customers to get help and share best practices. Like all official New Relic open source projects, there's a related Community topic in the New Relic Explorers Hub. You can find this project's topic/threads here:

https://discuss.newrelic.com/t/nr1-apdex-optimizer-nerdpack/92054

## Issues / enhancement requests

Issues and enhancement requests can be submitted in the [Issues tab of this repository](../../issues). Please search for and review the existing open issues before submitting a new issue.

# Contributing

Contributions are welcome and we encourage you to submit enhancements. Please review our [Contributors Guide](CONTRIBUTING.md) before you begin.

When you submit your pull request, you'll need to sign the CLA by using CLA-Assistant. If you'd like to execute our corporate CLA, or if you have any questions, please drop us an email at opensource+nr1-apdex-optimizer@newrelic.com.


docs/CHANGELOG.md
## [1.2.3](https://github.com/newrelic/nr1-apdex-optimizer/compare/v1.2.2...v1.2.3) (2020-06-05)


### Bug Fixes

* Update CI workflows ([#8](https://github.com/newrelic/nr1-apdex-optimizer/issues/8)) [generate new changelog via `fix`] ([e015eaa](https://github.com/newrelic/nr1-apdex-optimizer/commit/e015eaacbcda6cb20223821ec01468f3a3ddab4a))


nerdlets/nr1-apdex-optimizer-nerdlet/index.js
import React from 'react';
import {
  BlockText,
  HeadingText,
  PlatformStateContext,
  Stack,
  StackItem
} from 'nr1';
import AccountListSelect from './AccountListSelect';
import ApdexTableContainer from './ApdexTableContainer';

// https://docs.newrelic.com/docs/new-relic-programmable-platform-introduction

const LINK_COLOR = 'rgb(0, 121, 191)';

export default class ApdexOptimizer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      accountId: null
    };
    this.onAccountChange = this.onAccountChange.bind(this);
  }

  onAccountChange(accountId) {
    this.setState({
      accountId: accountId
    });
  }

  render() {
    return (
      <Stack
        directionType={Stack.DIRECTION_TYPE.VERTICAL}
        horizontalType={Stack.HORIZONTAL_TYPE.FILL}
        style={{ paddingLeft: 20, paddingRight: 20, width: '98%' }}
      >
        <StackItem>
          <AccountListSelect onAccountChange={this.onAccountChange} />
        </StackItem>
        <StackItem>
          <PlatformStateContext.Consumer>
            {platformState =>
              this.state.accountId ? (
                <ApdexTableContainer
                  accountId={this.state.accountId}
                  platformState={platformState}
                />
              ) : (
                <BlockText>&nbsp;</BlockText>
              )
            }
          </PlatformStateContext.Consumer>
        </StackItem>
        <StackItem>
          <HeadingText style={{ marginBottom: 8 }}>
            What should I set T to?
          </HeadingText>
          <BlockText type="paragraph" style={{ marginBottom: 8 }}>
            If you have an app that has been running for awhile in a steady
            state and you feel you have a good baseline for acceptable
            performance, you can start by setting your Apdex threshold to give
            you a baseline Apdex score of 0.95. So you’ll want to get the 90th
            percentile value and set that to Apdex T.
          </BlockText>
          <BlockText type="paragraph" style={{ marginBottom: 8 }}>
            The suggested thresholds above are based on the 90th percentile. It
            is recommended that the time picker is set to 7 days.
          </BlockText>
          <BlockText type="paragraph" style={{ marginBottom: 8 }}>
            Based on a mathematical{' '}
            <a
              target="_blank"
              rel="noopener noreferrer"
              href="https://blog.newrelic.com/product-news/how-to-choose-apdex-t/"
              style={{ color: LINK_COLOR }}
            >
              analysis
            </a>
            .
          </BlockText>
        </StackItem>
      </Stack>
    );
  }
}


nerdlets/nr1-apdex-optimizer-nerdlet/AccountListSelect.js
import React from 'react';
import PropTypes from 'prop-types';
import { AccountsQuery, Dropdown, DropdownItem, Spinner } from 'nr1';

export default class AccountListSelect extends React.Component {
  static propTypes = {
    onAccountChange: PropTypes.func.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      accountName: 'Select account...'
    };
    this.onAccountChange = this.onAccountChange.bind(this);
  }

  // Actions to perform on initial load
  componentDidMount() {
    this.queryAccounts();
  }

  onAccountChange(item) {
    this.setState({ accountName: item.name });
    this.props.onAccountChange(item.id);
  }

  async queryAccounts() {
    const { data } = await AccountsQuery.query();
    // Automatically load the account if there is only one
    if (data.length === 1) this.onAccountChange(data[0]);
  }

  render() {
    return (
      <AccountsQuery>
        {({ loading, error, data }) => {
          if (loading) {
            return <Spinner />;
          }

          if (error) {
            return 'Error!';
          }

          const dropDownitems = data.map(item => {
            return (
              <DropdownItem
                key={item.id}
                onClick={() => this.onAccountChange(item)}
              >
                {item.name}
              </DropdownItem>
            );
          });

          return (
            <Dropdown title={this.state.accountName}>{dropDownitems}</Dropdown>
          );
        }}
      </AccountsQuery>
    );
  }
}


nerdlets/nr1-apdex-optimizer-nerdlet/ApdexTableContainer.js
import React from 'react';
import PropTypes from 'prop-types';
import { NerdGraphQuery } from 'nr1';
import gql from 'graphql-tag';
import ApdexTable from './ApdexTable';

const SUGGESTED_APDEX_PERCENTILE = 90; // https://blog.newrelic.com/product-news/how-to-choose-apdex-t/

// Class used to encapsulate Apdex for a service
class ApdexRow {
  constructor(aName) {
    this.name = aName;
  }

  accountId = null;
  apmAppId = null;
  apmApdexT = null;
  apmApdexTHref = null;
  apmSuggestedApdexT = null;
  apmApdexScore = null;
  apmCount = null;
  apmErrorCount = null;
  browserAppId = null;
  browserApdexT = null;
  browserApdexTHref = null;
  browserSuggestedApdexT = null;
  browserApdexScore = null;
  browserCount = null;
  browserErrorCount = null;
}

export default class ApdexTableContainer extends React.Component {
  static propTypes = {
    accountId: PropTypes.number.isRequired,
    platformState: PropTypes.object.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      data: [],
      isLoading: true
    };
  }

  // Actions to perform on initial load
  componentDidMount() {
    this.fetchData();
  }

  // Actions to perform if the selected account changes
  componentDidUpdate(prevProps) {
    if (this.props !== prevProps) {
      // this.state.data = [];
      this.fetchData();
    }
  }

  async fetchData() {
    this.setState({ isLoading: true });
    const rowsMap = new Map();
    const response = await NerdGraphQuery.query(
      this.appsQuery(this.props.accountId)
    );
    const entities = response.data.actor.entitySearch.results.entities;
    for (let i = 0; i < entities.length; i++) {
      const appName = entities[i].name;
      if (!rowsMap.has(appName)) {
        const row = new ApdexRow(appName);
        const domain = entities[i].domain;
        row.accountId = this.props.accountId;
        if (domain === 'APM') {
          row.apmAppId = entities[i].applicationId;
          if (entities[i].settings)
            row.apmApdexT = entities[i].settings.apdexTarget;
          row.apmApdexTHref = `https://rpm.newrelic.com/accounts/${row.accountId}/applications/${row.apmAppId}/settings-application`;
        } else if (domain === 'BROWSER') {
          row.browserAppId = entities[i].applicationId;
          if (entities[i].settings)
            row.browserApdexT = entities[i].settings.apdexTarget;
          row.browserApdexTHref = `https://rpm.newrelic.com/accounts/${row.accountId}/browser/${row.apmAppId}/edit`;
        }
        rowsMap.set(appName, row);
      } else {
        const row = rowsMap.get(appName);
        const domain = entities[i].domain;
        if (domain === 'APM') {
          row.apmAppId = entities[i].applicationId;
          if (entities[i].settings)
            row.apmApdexT = entities[i].settings.apdexTarget;
          row.apmApdexTHref = `https://rpm.newrelic.com/accounts/${row.accountId}/applications/${row.apmAppId}/settings-application`;
        } else if (domain === 'BROWSER') {
          row.browserAppId = entities[i].applicationId;
          if (entities[i].settings)
            row.browserApdexT = entities[i].settings.apdexTarget;
          row.browserApdexTHref = `https://rpm.newrelic.com/accounts/${row.accountId}/browser/${row.apmAppId}/edit`;
        }
      }
    }
    await this.fetchStats(rowsMap);
    await this.fetchErrors(rowsMap);
    await this.fetchSuggestedApdexT(rowsMap, 'Transaction');
    await this.fetchSuggestedApdexT(rowsMap, 'PageView');

    const data = await this.getData(rowsMap);

    this.setState({
      data,
      isLoading: false
    });
  }

  // NerdGraph grapql (nrql) query of suggested Apdex Threshold based on percentile and other supplied parameters
  suggestedApdexTQuery(accountId, percentile, eventType) {
    const { duration } = this.props.platformState.timeRange;
    const since = `SINCE ${duration / 1000 / 60} MINUTES AGO`;
    const where =
      eventType === 'Transaction' ? "WHERE transactionType = 'Web'" : ''; // Only Web Transactions
    return {
      query: gql`{
                actor {
                    account(id: ${accountId}) {
                        nrql(query: "SELECT percentile(duration, ${percentile}) FROM ${eventType} ${where} ${since} FACET appName LIMIT MAX") {
                            results
                        }
                    }
                }
            }`
    };
  }

  // NerdGraph grapql (nrql) query of Apdex Score and volume based on percentile and other supplied parameters
  statsQuery(accountId, threshold, eventType) {
    const { duration } = this.props.platformState.timeRange;
    const since = `SINCE ${duration / 1000 / 60} MINUTES AGO`;
    return {
      query: gql`{
                actor {
                    account(id: ${accountId}) {
                        nrql(query: "SELECT apdex(duration, t: ${threshold}), count(*) AS 'Throughput' FROM ${eventType} ${since} FACET appName LIMIT MAX") {
                          results
                        }
                    }
                }
            }`
    };
  }

  // NerdGraph grapql (nrql) query of errors based on supplied parameters
  errorsQuery(accountId, eventType) {
    const { duration } = this.props.platformState.timeRange;
    const since = `SINCE ${duration / 1000 / 60} MINUTES AGO`;
    return {
      query: gql`{
                actor {
                    account(id: ${accountId}) {
                        nrql(query: "SELECT count(*) FROM ${eventType} ${since} FACET appName LIMIT MAX") {
                          results
                        }
                    }
                }
            }`
    };
  }

  // NerdGraph graphql query of entities in the APM and BROWSER domains
  appsQuery(accountId) {
    return {
      query: gql`{
                actor {
                    entitySearch(query: "accountId = '${accountId}' AND domain IN ('APM', 'BROWSER')") {
                        results {
                            entities {
                                name
                                domain
                                entityType
                                ... on BrowserApplicationEntityOutline {
                                    accountId
                                    alertSeverity
                                    settings {
                                        apdexTarget
                                    }
                                    reporting
                                    applicationId
                                }
                                ... on ApmApplicationEntityOutline {
                                    accountId
                                    alertSeverity
                                    settings {
                                        apdexTarget
                                    }
                                    apmSummary {
                                        apdexScore
                                    }
                                    reporting
                                    applicationId
                                }
                            }
                        nextCursor
                        }
                    }
                }
            }`
    };
  }

  // Convert retrieved rows to ReactTable compatible data (array of Objects)
  async getData(rowsMap) {
    const data = [];
    rowsMap.forEach(value => {
      let apmErrorCount;
      if (value.apmErrorCount) {
        apmErrorCount = value.apmErrorCount;
      } else {
        apmErrorCount = value.apmApdexT ? 0 : '';
      }
      let browserErrorCount;
      if (value.browserErrorCount) {
        browserErrorCount = value.browserErrorCount;
      } else {
        browserErrorCount = value.browserApdexT ? 0 : '';
      }
      data.push({
        name: value.name,
        accountId: value.accountId,
        apmApdexT: value.apmApdexT,
        apmApdexTHref: value.apmApdexTHref,
        apmSuggestedApdexT:
          value.apmSuggestedApdexT && value.apmSuggestedApdexT.toFixed(3),
        apmApdexScore: value.apmApdexScore && value.apmApdexScore.toFixed(2),
        apmCount: value.apmCount,
        apmErrorCount: apmErrorCount,
        browserApdexT: value.browserApdexT,
        browserApdexTHref: value.browserApdexTHref,
        browserSuggestedApdexT:
          value.browserSuggestedApdexT &&
          value.browserSuggestedApdexT.toFixed(3),
        browserApdexScore:
          value.browserApdexScore && value.browserApdexScore.toFixed(2),
        browserCount: value.browserCount,
        browserErrorCount: browserErrorCount
      });
    });
    return data;
  }

  async fetchStats(rowsMap) {
    let response = await NerdGraphQuery.query(
      this.statsQuery(this.props.accountId, 0.5, 'Transaction')
    );
    let results = response.data.actor.account.nrql.results;
    for (let i = 0; i < results.length; i++) {
      const appName = results[i].appName;
      const row = rowsMap.get(appName);
      if (row) {
        row.apmCount = results[i].count;
        row.apmApdexScore = results[i].score;
      }
    }
    response = await NerdGraphQuery.query(
      this.statsQuery(this.props.accountId, 7.0, 'PageView')
    );
    results = response.data.actor.account.nrql.results;
    for (let i = 0; i < results.length; i++) {
      const appName = results[i].appName;
      const row = rowsMap.get(appName);
      if (row) {
        row.browserCount = results[i].count;
        row.browserApdexScore = results[i].score;
      }
    }
  }

  async fetchErrors(rowsMap) {
    let response = await NerdGraphQuery.query(
      this.errorsQuery(this.props.accountId, 'TransactionError')
    );
    let results = response.data.actor.account.nrql.results;
    for (let i = 0; i < results.length; i++) {
      const appName = results[i].appName;
      const row = rowsMap.get(appName);
      if (row) {
        row.apmErrorCount = results[i].count;
      }
    }
    response = await NerdGraphQuery.query(
      this.errorsQuery(this.props.accountId, 'JavaScriptError')
    );
    results = response.data.actor.account.nrql.results;
    for (let i = 0; i < results.length; i++) {
      const appName = results[i].appName;
      const row = rowsMap.get(appName);
      if (row) {
        row.browserErrorCount = results[i].count;
      }
    }
  }

  async fetchSuggestedApdexT(rowsMap, eventType) {
    const { data } = await NerdGraphQuery.query(
      this.suggestedApdexTQuery(
        this.props.accountId,
        SUGGESTED_APDEX_PERCENTILE,
        eventType
      )
    );
    const results = data.actor.account.nrql.results;
    for (let i = 0; i < results.length; i++) {
      const appName = results[i].appName;
      const row = rowsMap.get(appName);
      if (row) {
        const suggestedApdexT =
          results[i]['percentile.duration'][SUGGESTED_APDEX_PERCENTILE];
        eventType === 'Transaction'
          ? (row.apmSuggestedApdexT = suggestedApdexT)
          : (row.browserSuggestedApdexT = suggestedApdexT);
      }
    }
  }

  render() {
    return (
      <ApdexTable data={this.state.data} isLoading={this.state.isLoading} />
    );
  }
}


nerdlets/nr1-apdex-optimizer-nerdlet/ApdexTable.js
import React from 'react';
import PropTypes from 'prop-types';
import ReactTable, { ReactTableDefaults } from 'react-table';
import 'react-table/react-table.css';
import './styles.scss';

const DEFAULT_APM_APDEX_T = 0.5;
const DEFAULT_BROWSER_APDEX_T = 7;

// Existing UI Colors
const CRITICAL_COLOR = 'rgb(191, 0, 22)';
// const WARNING_COLOR = 'rgb(191, 150, 0)';
const NORMAL_COLOR = 'rgb(17, 166, 0)';
const LINK_COLOR = 'rgb(0, 121, 191)';

const columnDefaults = {
  ...ReactTableDefaults.column,
  headerClassName: 'wordwrap',
  headerStyle: { textAlign: 'center' }
};
const columns = [
  {
    Header: 'Application Name',
    accessor: 'name'
  },
  {
    Header: 'APM Transactions',
    accessor: 'apmCount',
    className: 'right'
  },
  {
    Header: 'APM Transaction Errors',
    accessor: 'apmErrorCount',
    className: 'right'
  },
  {
    Header: 'Configured APM ApdexT',
    accessor: 'apmApdexT',
    Cell: cellInfo =>
      cellInfo.original.apmApdexT && (
        <a
          target="_blank"
          rel="noopener noreferrer"
          href={cellInfo.original.apmApdexTHref}
          title="Click to launch APM app settings page in a new tab"
          style={{ color: LINK_COLOR }}
        >
          {cellInfo.original.apmApdexT}
        </a>
      ),
    className: 'right'
  },
  {
    Header: 'APM Apdex Score [t:0.5]',
    accessor: 'apmApdexScore',
    className: 'right'
  },
  {
    Header: 'Suggested APM ApdexT',
    accessor: 'apmSuggestedApdexT',
    getProps: (state, rowInfo) => {
      return {
        style: {
          color:
            rowInfo && rowInfo.row.apmSuggestedApdexT <= DEFAULT_APM_APDEX_T
              ? NORMAL_COLOR
              : CRITICAL_COLOR
        }
      };
    },
    className: 'right'
  },
  {
    Header: 'Browser Page Views',
    accessor: 'browserCount',
    className: 'right'
  },
  {
    Header: 'JavaScript Errors',
    accessor: 'browserErrorCount',
    className: 'right'
  },
  {
    Header: 'Configured Browser ApdexT',
    accessor: 'browserApdexT',
    Cell: cellInfo =>
      cellInfo.original.browserApdexT && (
        <a
          target="_blank"
          rel="noopener noreferrer"
          href={cellInfo.original.browserApdexTHref}
          title="Click to launch Browser app settings page in a new tab"
          style={{ color: LINK_COLOR }}
        >
          {cellInfo.original.browserApdexT}
        </a>
      ),
    className: 'right'
  },
  {
    Header: 'Browser Apdex Score [t:7.0]',
    accessor: 'browserApdexScore',
    className: 'right'
  },
  {
    Header: 'Suggested Browser ApdexT',
    accessor: 'browserSuggestedApdexT',
    getProps: (state, rowInfo) => {
      return {
        style: {
          color:
            rowInfo &&
            rowInfo.row.browserSuggestedApdexT <= DEFAULT_BROWSER_APDEX_T
              ? NORMAL_COLOR
              : CRITICAL_COLOR
        }
      };
    },
    className: 'right'
  }
];

export default class ApdexTable extends React.Component {
  static propTypes = {
    data: PropTypes.array.isRequired,
    isLoading: PropTypes.bool.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      search: ''
    };
  }

  render() {
    let data = this.props.data;
    if (this.state.search) {
      data = data.filter(row => {
        return (
          row.name.includes(this.state.search) ||
          String(row.age).includes(this.state.search) ||
          String(row.apmApdexT).includes(this.state.search) ||
          String(row.browserApdexT).includes(this.state.search) ||
          String(row.apmApdexScore).includes(this.state.search) ||
          String(row.browserApdexScore).includes(this.state.search)
        );
      });
    }

    return (
      <div>
        Search:{' '}
        <input
          value={this.state.search}
          onChange={e => this.setState({ search: e.target.value })}
          style={{ border: '1px solid gray', width: '20%' }}
        />
        <p>&nbsp;</p>
        <ReactTable
          data={data}
          column={columnDefaults}
          columns={columns}
          className="-striped -highlight"
          minRows={1}
          defaultPageSize={10}
          pageSizeOptions={[10, 25, 50, 100]}
          loading={this.props.isLoading}
        />
      </div>
    );
  }
}
